File: express-backend/routes/workspace.js
import { Router } from "express";
import { pool } from "../db.js";

const router = Router();

// Self-creating, same convention as the officers table in addOfficer.js —
// no manual migration step needed.
// person_id is VARCHAR to match persons.person_id (e.g. "P001").
async function ensureWorkspaceTables() {
  await pool.query(`
    CREATE TABLE IF NOT EXISTS officer_pinned_criminal (
      officer_id  INTEGER PRIMARY KEY REFERENCES officers(officer_id) ON DELETE CASCADE,
      person_id   VARCHAR(20) NOT NULL REFERENCES persons(person_id) ON DELETE CASCADE,
      pinned_at   TIMESTAMPTZ NOT NULL DEFAULT now()
    );
  `);
  await pool.query(`
    CREATE TABLE IF NOT EXISTS officer_working_list (
      officer_id  INTEGER NOT NULL REFERENCES officers(officer_id) ON DELETE CASCADE,
      person_id   VARCHAR(20) NOT NULL REFERENCES persons(person_id) ON DELETE CASCADE,
      added_at    TIMESTAMPTZ NOT NULL DEFAULT now(),
      PRIMARY KEY (officer_id, person_id)
    );
  `);
}
ensureWorkspaceTables().catch((err) => console.error("Failed to ensure workspace tables", err));

// Joins against persons/cases/crime_types instead of storing a name+tags
// snapshot, so the list always reflects current data (unlike the old
// localStorage snapshot, which could go stale).
const LIST_ITEMS_SQL = `
  SELECT p.person_id, p.name,
         array_remove(array_agg(DISTINCT ct.crime_name), NULL) AS crime_tags
  FROM officer_working_list w
  JOIN persons p ON p.person_id = w.person_id
  LEFT JOIN cases c ON c.person_id = p.person_id
  LEFT JOIN crime_types ct ON ct.crime_id = c.crime_id
  WHERE w.officer_id = $1
  GROUP BY p.person_id, w.added_at
  ORDER BY w.added_at ASC
`;

// GET /api/workspace — this officer's pinned criminal + working list
router.get("/", async (req, res) => {
  try {
    const officerId = req.officer.officerId;
    const [{ rows: pinnedRows }, { rows: listRows }] = await Promise.all([
      pool.query(`SELECT person_id FROM officer_pinned_criminal WHERE officer_id = $1`, [officerId]),
      pool.query(LIST_ITEMS_SQL, [officerId]),
    ]);

    res.json({
      pinnedId: pinnedRows[0]?.person_id ?? null,
      workingList: listRows.map((r) => ({
        id: r.person_id,
        name: r.name,
        crimeTags: r.crime_tags ? r.crime_tags.filter(Boolean) : [],
      })),
    });
  } catch (err) {
    console.error("GET /api/workspace failed", err);
    res.status(500).json({ error: "Failed to load workspace" });
  }
});

// PUT /api/workspace/pin { personId } — pin replaces whatever this officer had pinned
router.put("/pin", async (req, res) => {
  const { personId } = req.body || {};
  if (!personId) return res.status(400).json({ error: "personId is required" });

  try {
    await pool.query(
      `INSERT INTO officer_pinned_criminal (officer_id, person_id, pinned_at)
       VALUES ($1, $2, now())
       ON CONFLICT (officer_id) DO UPDATE SET person_id = EXCLUDED.person_id, pinned_at = now()`,
      [req.officer.officerId, personId]
    );
    res.json({ pinnedId: personId });
  } catch (err) {
    console.error("PUT /api/workspace/pin failed", err);
    res.status(500).json({ error: "Failed to pin criminal" });
  }
});

// DELETE /api/workspace/pin — unpin for this officer
router.delete("/pin", async (req, res) => {
  try {
    await pool.query(`DELETE FROM officer_pinned_criminal WHERE officer_id = $1`, [req.officer.officerId]);
    res.json({ pinnedId: null });
  } catch (err) {
    console.error("DELETE /api/workspace/pin failed", err);
    res.status(500).json({ error: "Failed to unpin criminal" });
  }
});

// POST /api/workspace/list { personId } — add to this officer's working list
router.post("/list", async (req, res) => {
  const { personId } = req.body || {};
  if (!personId) return res.status(400).json({ error: "personId is required" });

  try {
    await pool.query(
      `INSERT INTO officer_working_list (officer_id, person_id)
       VALUES ($1, $2)
       ON CONFLICT (officer_id, person_id) DO NOTHING`,
      [req.officer.officerId, personId]
    );
    res.status(201).json({ ok: true });
  } catch (err) {
    console.error("POST /api/workspace/list failed", err);
    res.status(500).json({ error: "Failed to add to list" });
  }
});

// DELETE /api/workspace/list/:personId — remove from this officer's working list
router.delete("/list/:personId", async (req, res) => {
  try {
    await pool.query(
      `DELETE FROM officer_working_list WHERE officer_id = $1 AND person_id = $2`,
      [req.officer.officerId, req.params.personId]
    );
    res.json({ ok: true });
  } catch (err) {
    console.error("DELETE /api/workspace/list failed", err);
    res.status(500).json({ error: "Failed to remove from list" });
  }
});

export default router;

File: backend/routes/workspace.py
import asyncio

from fastapi import APIRouter, Depends, Request

from ..errors import APIError, api_errors
from ..models import PersonBody
from ..security import require_auth

router = APIRouter(prefix="/api/workspace", tags=["Workspace"])
LIST_ITEMS_SQL = """
  SELECT p.person_id, p.name,
         array_remove(array_agg(DISTINCT ct.crime_name), NULL) AS crime_tags
  FROM officer_working_list w
  JOIN persons p ON p.person_id = w.person_id
  LEFT JOIN case_people cp ON cp.person_id = p.person_id
  LEFT JOIN cases c ON c.case_id = cp.case_id
  LEFT JOIN crime_types ct ON ct.crime_id = c.crime_id
  WHERE w.officer_id = %s
  GROUP BY p.person_id, w.added_at
  ORDER BY w.added_at ASC
"""


@router.get("/", include_in_schema=False)
@router.get("")
async def get_workspace(request: Request, officer=Depends(require_auth)):
    # Scope both the pin and working list to the signed-in officer.
    with api_errors("Failed to load workspace"):
        pinned, rows = await asyncio.gather(
            request.app.state.db.query(
                "SELECT person_id FROM officer_pinned_criminal WHERE officer_id = %s",
                (officer["officerId"],),
            ),
            request.app.state.db.query(LIST_ITEMS_SQL, (officer["officerId"],)),
        )
        return {
            "pinnedId": pinned[0]["person_id"] if pinned else None,
            "workingList": [
                {
                    "id": row["person_id"],
                    "name": row["name"],
                    "crimeTags": list(filter(None, row.get("crime_tags") or [])),
                }
                for row in rows
            ],
        }


@router.put("/pin")
async def pin(request: Request, body: PersonBody | None = None, officer=Depends(require_auth)):
    if body is None or not body.personId:
        raise APIError("personId is required", 400)
    with api_errors("Failed to pin criminal"):
        # Each officer has one pin; pinning again replaces it.
        await request.app.state.db.query(
            """INSERT INTO officer_pinned_criminal (officer_id, person_id, pinned_at)
               VALUES (%s, %s, now())
               ON CONFLICT (officer_id) DO UPDATE
               SET person_id = EXCLUDED.person_id, pinned_at = now()""",
            (officer["officerId"], body.personId),
        )
        return {"pinnedId": body.personId}


@router.delete("/pin")
async def unpin(request: Request, officer=Depends(require_auth)):
    with api_errors("Failed to unpin criminal"):
        await request.app.state.db.query(
            "DELETE FROM officer_pinned_criminal WHERE officer_id = %s",
            (officer["officerId"],),
        )
        return {"pinnedId": None}


@router.post("/list", status_code=201)
async def add_to_list(
    request: Request, body: PersonBody | None = None, officer=Depends(require_auth)
):
    if body is None or not body.personId:
        raise APIError("personId is required", 400)
    with api_errors("Failed to add to list"):
        # Repeated additions leave the existing list entry unchanged.
        await request.app.state.db.query(
            """INSERT INTO officer_working_list (officer_id, person_id) VALUES (%s, %s)
               ON CONFLICT (officer_id, person_id) DO NOTHING""",
            (officer["officerId"], body.personId),
        )
        return {"ok": True}


@router.delete("/list/{person_id}")
async def remove_from_list(person_id: str, request: Request, officer=Depends(require_auth)):
    with api_errors("Failed to remove from list"):
        await request.app.state.db.query(
            "DELETE FROM officer_working_list WHERE officer_id = %s AND person_id = %s",
            (officer["officerId"], person_id),
        )
        return {"ok": True}
